subsys/bluetooth/host/smp.c

```diff
@@ -647,7 +647,10 @@ static bool update_keys_check(struct bt_smp *smp, struct bt_keys *keys)
 	}
 
 	if ((keys->flags & BT_KEYS_AUTHENTICATED) &&
-	     smp->method == JUST_WORKS) {
+		smp->method == JUST_WORKS) {
+		if (IS_ENABLED(CONFIG_BT_SMP_ALLOW_UNAUTH_OVERWRITE_AUTHENTICATED)) {
```

Comment:
As an alternative, we could consider making this (and possibly the others) a application callback. 

With this PR we now have `BT_SMP_ALLOW_UNAUTH_OVERWRITE`, `BT_ID_ALLOW_UNAUTH_OVERWRITE` and `BT_SMP_ALLOW_UNAUTH_OVERWRITE_AUTHENTICATED`. Not sure if we end up with more in the future, but we end up several combinations that are mostly untested by Zephyr. 

Furthermore, an application may want this to be done conditionally, e.g. it may allow going from level 3 to 2, but not level 4 to 2, or done with devices that has a specific service or from a specific vendor.